core/kotlinx-coroutines-test/src/TestCoroutineDispatcher.kt

```diff
@@ -0,0 +1,294 @@
+ * methods on [DelayController]
+ *
+ * When swiched to lazy execution any coroutines started via [launch] or [async] will
+ * not execute until a call to [DelayController.runCurrent] or the virtual clock-time has been advanced via one of the
+ * methods on [DelayController].
+ *
+ * @see DelayController
+ */
+@ExperimentalCoroutinesApi
+class TestCoroutineDispatcher:
+        CoroutineDispatcher(),
+        Delay,
+        DelayController {
+
+    private var dispatchImmediately = true
+        set(value) {
+            field = value
+            if (value) {
+                // there may already be tasks from setup code we need to run
+                advanceUntilIdle()
+            }
+        }
+
+    // The ordered queue for the runnable tasks.
+    private val queue = ThreadSafeHeap<TimedRunnable>()
+
+    // The per-scheduler global order counter.
+    private var counter = AtomicLong(0)
+
+    // Storing time in nanoseconds internally.
+    private var time = AtomicLong(0)
+
+    override fun dispatch(context: CoroutineContext, block: Runnable) {
+        if (dispatchImmediately) {
+            block.run()
+        } else {
+            post(block)
+        }
+    }
+
+    override fun scheduleResumeAfterDelay(timeMillis: Long, continuation: CancellableContinuation<Unit>) {
+        postDelayed(CancellableContinuationRunnable(continuation) { resumeUndispatched(Unit) }, timeMillis)
+    }
+
+    override fun invokeOnTimeout(timeMillis: Long, block: Runnable): DisposableHandle {
+        val node = postDelayed(block, timeMillis)
+        return object : DisposableHandle {
+            override fun dispose() {
+                queue.remove(node)
+            }
+        }
+    }
+
+    override fun toString(): String {
+        return "TestCoroutineDispatcher[currentTime=${time}ms, queued=${queue.size}]"
+    }
+
+    private fun post(block: Runnable) =
+            queue.addLast(TimedRunnable(block, counter.getAndIncrement()))
+
+    private fun postDelayed(block: Runnable, delayTime: Long) =
+            TimedRunnable(block, counter.getAndIncrement(), time.get() + delayTime)
+                    .also {
+                        queue.addLast(it)
+                    }
+
+
+    private fun doActionsUntil(targetTime: Long) {
+        while (true) {
+            val current = queue.removeFirstIf { it.time <= targetTime } ?: break
+            // If the scheduled time is 0 (immediate) use current virtual time
+            time.getAndAccumulate(current.time) { currentValue: Long, proposedValue: Long ->
+                if (proposedValue != 0L) {
+                    proposedValue
+                } else {
+                    currentValue
+                }
+            }
+            current.run()
+        }
+    }
+
+    override fun currentTime() = time.get()
+
+    override fun advanceTimeBy(delayTimeMillis: Long): Long {
+        val oldTime = time.get()
+        advanceUntilTime(oldTime + delayTimeMillis)
```

Comment:
Be sure to clamp the sum `oldTime + delayTimeMillis` against Long.MAX_VALUE so that this sum would not accidentally become a negative value.

Reply:
I'm taking care of it